File: frontend/src/services/project.ts
import { api } from "./api";

export interface Project {
  id: string;
  workspace_id: string;
  name: string;
  slug: string;
  description: string | null;
  project_url: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface ProjectCreate {
  name: string;
  description?: string;
  project_url?: string;
}

export interface ProjectUpdate {
  name?: string;
  description?: string;
  project_url?: string;
}

export async function getProjects(
  workspaceId: string,
): Promise<Project[]> {
  const response = await api.get<Project[]>(
    `/api/v1/workspaces/${workspaceId}/projects`,
  );

  return response.data;
}

export async function getProject(
  workspaceId: string,
  projectId: string,
): Promise<Project> {
  const response = await api.get<Project>(
    `/api/v1/workspaces/${workspaceId}/projects/${projectId}`,
  );

  return response.data;
}

export async function createProject(
  workspaceId: string,
  data: ProjectCreate,
): Promise<Project> {
  const response = await api.post<Project>(
    `/api/v1/workspaces/${workspaceId}/projects`,
    data,
  );

  return response.data;
}

export async function updateProject(
  workspaceId: string,
  projectId: string,
  data: ProjectUpdate,
): Promise<Project> {
  const response = await api.patch<Project>(
    `/api/v1/workspaces/${workspaceId}/projects/${projectId}`,
    data,
  );

  return response.data;
}

export async function deleteProject(
  workspaceId: string,
  projectId: string,
): Promise<void> {
  await api.delete(
    `/api/v1/workspaces/${workspaceId}/projects/${projectId}`,
  );
}

File: backend/app/api/routes/project.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.dependencies import get_current_user_id
from app.api.permission import require_workspace_role
from app.core.database import get_db
from app.models.workspace_membership import WorkspaceMembership
from app.schemas.project import (
    ProjectCreate,
    ProjectResponse,
    ProjectUpdate,
)
from app.services.project import (
    create_project,
    delete_project,
    get_project,
    get_projects,
    update_project,
)


router = APIRouter(
    prefix="/api/v1/workspaces",
    tags=["Projects"],
)


def serialize_project(project):
    return ProjectResponse(
        id=str(project.id),
        workspace_id=str(project.workspace_id),
        name=project.name,
        slug=project.slug,
        description=project.description,
        project_url=getattr(project, "project_url", None),
        created_by=str(project.created_by),
        created_at=project.created_at,
        updated_at=project.updated_at,
    )


@router.post(
    "/{workspace_id}/projects",
    response_model=ProjectResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_workspace_project(
    workspace_id: UUID,
    data: ProjectCreate,
    db: Session = Depends(get_db),
    current_user_id: str = Depends(get_current_user_id),
    membership: WorkspaceMembership = Depends(
        require_workspace_role("owner", "admin", "member")
    ),
):
    project = create_project(
        db=db,
        workspace_id=str(workspace_id),
        user_id=current_user_id,
        data=data,
    )

    return serialize_project(project)


@router.get(
    "/{workspace_id}/projects",
    response_model=list[ProjectResponse],
)
def list_workspace_projects(
    workspace_id: UUID,
    db: Session = Depends(get_db),
    membership: WorkspaceMembership = Depends(
        require_workspace_role("owner", "admin", "member")
    ),
):
    projects = get_projects(
        db,
        str(workspace_id),
    )

    return [
        serialize_project(project)
        for project in projects
    ]


@router.get(
    "/{workspace_id}/projects/{project_id}",
    response_model=ProjectResponse,
)
def get_workspace_project(
    workspace_id: UUID,
    project_id: UUID,
    db: Session = Depends(get_db),
    membership: WorkspaceMembership = Depends(
        require_workspace_role("owner", "admin", "member")
    ),
):
    project = get_project(
        db,
        str(workspace_id),
        str(project_id),
    )

    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found",
        )

    return serialize_project(project)


@router.patch(
    "/{workspace_id}/projects/{project_id}",
    response_model=ProjectResponse,
)
def update_workspace_project(
    workspace_id: UUID,
    project_id: UUID,
    data: ProjectUpdate,
    db: Session = Depends(get_db),
    membership: WorkspaceMembership = Depends(
        require_workspace_role("owner", "admin", "member")
    ),
):
    project = get_project(
        db,
        str(workspace_id),
        str(project_id),
    )

    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found",
        )

    project = update_project(
        db,
        project,
        data,
    )

    return serialize_project(project)


@router.delete(
    "/{workspace_id}/projects/{project_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_workspace_project(
    workspace_id: UUID,
    project_id: UUID,
    db: Session = Depends(get_db),
    membership: WorkspaceMembership = Depends(
        require_workspace_role("owner")
    ),
):
    project = get_project(
        db,
        str(workspace_id),
        str(project_id),
    )

    if not project:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found",
        )

    delete_project(db, project)

    return None